### PR TITLE
Add support of dynamic graphs and code for ROLAND.

### PR DESCRIPTION
**Pull request by TianyuDu**

### CODE REVIEW

File: ROLAND_README.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+To explore training result:
+```bash
+cd ./run
+tensorboard --logdir=./runs_live_update --port=6006
+```
+**WARNING** The x-axis of plots in tensorboard is **not** epochs, they are snapshot IDs (e.g., the $i^{th}$ day or the $i^{th}$ week) instead.
+
+## Examples on Heterogenous Graph Snapshots
+```bash
+Under development.
```
Comment (JiaxuanYou) [inline ROLAND_README.md:75]: You can start writing the example, without worrying about heterogenous layer yet.

File: graphgym/contrib/config/roland.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    # ----------------------------------------------------------------------- #
+
+    cfg.metric = CN()
+    # How many negative edges for each node to compute rank-based evaluation
+    # metrics such as MRR and recall at K.
+    # E.g., if multiplier = 1000 and a node has 3 positive edges, then we
+    # compute the MRR using 1000 randomly generated negative edges
+    # + 3 existing positive edges.
+    # Use 100 ~ 1000 for fast and reliable results.
+    cfg.metric.mrr_num_negative_edges = 1000
```
Comment (JiaxuanYou) [inline graphgym/contrib/config/roland.py:184]: How can we set negative edges as all the edges?

File: graphgym/contrib/loader/roland_as.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+def load_generic_dataset(format, name, dataset_dir):
+    if format == 'as':
+        all_files = [x for x in sorted(os.listdir(dataset_dir))
+                     if (x.startswith('as') and x.endswith('.txt'))]
+        assert len(all_files) == 733
+        assert all(x.endswith('.txt') for x in all_files)
+
+        edge_index_lst, edge_time_lst = list(), list()
+        all_files = sorted(all_files)
+        # if cfg.train.mode in ['baseline', 'baseline_v2', 'live_update_fixed_split']:
```
Comment (JiaxuanYou) [inline graphgym/contrib/loader/roland_as.py:77]: clean

File: graphgym/contrib/loader/roland_as.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
+"""
```
Comment (JiaxuanYou) [inline graphgym/contrib/loader/roland_as.py:1]: We have merged `roland_xx` to a single file called `roland`, so I think this should be removed?

File: graphgym/contrib/loader/roland_btc.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
+"""
```
Comment (JiaxuanYou) [inline graphgym/contrib/loader/roland_btc.py:1]: We have merged `roland_xx` to a single file called `roland`, so I think this should be removed?

File: graphgym/contrib/loader/roland_reddit.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
+import os
```
Comment (JiaxuanYou) [inline graphgym/contrib/loader/roland_reddit.py:1]: We have merged `roland_xx` to a single file called `roland`, so I think this should be removed?

File: graphgym/contrib/loader/roland_ucimsg.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
+"""
```
Comment (JiaxuanYou) [inline graphgym/contrib/loader/roland_ucimsg.py:1]: We have merged `roland_xx` to a single file called `roland`, so I think this should be removed?

File: graphgym/contrib/train/train_utils.py [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+
+    # counts 1 + how many negative edge from src has higher score than p_best.
+    # if there's no such negative edge, rank is 1.
+    rank_by_user = compare.sum(axis=1) + 1  # (num_users,)
+    assert rank_by_user.shape == (num_users,)
+
+    mrr = float(torch.mean(1 / rank_by_user))
+    return mrr
+
+# TODO: get recall at k back.
```
Comment (JiaxuanYou) [inline graphgym/contrib/train/train_utils.py:271]: clean

File: graphgym/models/update.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        X = batch.node_feature
+        Z = self.GRU_Z(torch.cat([X, H_prev], dim=1))
+        R = self.GRU_R(torch.cat([X, H_prev], dim=1))
+        H_tilde = self.GRU_H_Tilde(torch.cat([X, R * H_prev], dim=1))
+        H_gru = Z * H_prev + (1 - Z) * H_tilde
+        batch.node_states[self.layer_id] = H_gru
+        return batch
+
+
+# class MaskedGRUUpdater(nn.Module):
```
Comment (JiaxuanYou) [inline graphgym/models/update.py:101]: clean

**Review by JiaxuanYou — COMMENTED** [review]
Please see the comments

File: graphgym/contrib/config/roland.py (hunk 2 above)
Comment (TianyuDu) [reply graphgym/contrib/config/roland.py:184]: We can set `cfg.metric.mrr_method = 'all'` to forcely enable it, but typically we don't use all negative edges since it's super slow.
